DOC-9205 - KV Operations dev-reorg (#2275)
couchbase/docs-server: modules/guides/pages/deleting-data.adoc

= Deleting Data
:description: How to delete documents in Couchbase.
:page-pagination: full 
:page-topic-type: guide
:tabs:
:github: Click the GitHub button icon:github[] to view this code in context.

include::partial$example-attributes.adoc[]

[abstract]
{description}

> 

Reviewer: For consistency

```suggestion

== Introduction

```

> In situations where data is no longer needed, Couchbase Server provides a `remove` operation to delete a document from the database permanently. 

Reviewer: Remove markup from `remove` thoughout this page where it's referring to the generic term for the operation, rather than a specific command or option

> 

Reviewer: Include links to the client tools and warning about altering your data

```suggestion

include::partial$clients.adoc[tag=refs]

include::partial$data-warning.adoc[]

```

== Deleting a Document

To delete a document from the database perform a `remove` operation.

> [{tabs}]

Reviewer: Add links to reference docs for every tab.

----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundException` error.

{github}
--

Java::
+ 
--
Use the `remove()` method to delete a document from the database.

{kv-delete-example}

[source,java]
----
include::java-sdk:hello-world:example$KvHelloWorldScoped.java[tag=kv-remove,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundException` error.

{github}
--

Node.js::
+ 
--
Use the `remove()` function to delete a document from the database.

{kv-delete-example}

[source,nodejs]
----
include::nodejs-sdk:hello-world:example$kv-hello-world-scoped.js[tag=kv-remove,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundError` error.

{github}
--

Python::
+ 
--
Use the `remove()` function to delete a document from the database.

{kv-delete-example}

[source,python]
----
include::python-sdk:hello-world:example$kv_hello_world_scoped.py[tag=kv-remove,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundException` error.

{github}
--
====

== Deleting a Sub-Document

> To delete a specific field within a document you can perform a _Sub-Document_ `remove` operation.

Reviewer: Tighten up wording to make it more procedural and less descriptive.

```suggestion
To delete a specific field within a document, perform a _Sub-Document_ remove operation.
```

[{tabs}]
====
cbc-subdoc::
+ 
--
. Connect to the `cbc-subdoc` interactive shell.

> . Use the `remove` command to delete a field from a document.
> 	Pass the field to remove with the `--path` argument.

Reviewer: Could be two steps if you want?

```suggestion
. Use the `remove` command to delete a field from a document.

. Pass the field to remove with the `--path` argument.
```

{kv-subdoc-delete-example}

[source,console]
----
cbc-subdoc -u Administrator -P password -U couchbase://localhost/travel-sample
subdoc> remove hotel-123 --path url
----

.Result
[source,console]
----
hotel-123          CAS=0x16be2f11c6040000
0. Size=0, RC=LCB_SUCCESS (0)
----

NOTE: If the path cannot be found, `cbc-subdoc` will return a `LCB_ERR_SUBDOC_PATH_NOT_FOUND` error. 
--

.NET::
+ 
--

> . To remove a specific field inside a document, call the `MutateInAsync()` method, which takes a document ID and an IEnumerable containing `MutateInSpec` objects. 
> `MutateInSpec` represents the sub-operation to be performed within the lookup.
> 
> . A `MutateInResult` object is returned containing the result and metadata relevant to the sub-document update operation.

Reviewer: Tighten up wording to make it more procedural and less descriptive.

```suggestion
. Call the `MutateInAsync()` method, which takes a document ID and an IEnumerable containing `MutateInSpec` objects.
 
. Use a `MutateInSpec` object to specify the sub-operation to be performed within the lookup.

A `MutateInResult` object is returned, containing the result and metadata relevant to the sub-document update operation.
```

{kv-subdoc-delete-example}

[source,csharp]
----
include::dotnet-sdk:hello-world:example$KvHelloWorldScoped.csx[tag=kv-remove-subdoc,indent=0]
----

NOTE: If the path doesn't exist, the SDK will return a `PathNotFoundException` error.

{github}
--

Java::
+ 
--

> . To remove a specific field inside a document, call the `mutateIn()` method, which takes a document ID and an array of `MutateInSpec` objects. `MutateInSpec` represents the sub-operation to be performed within the lookup.
> 
> . A `MutateInResult` object is returned containing the result and metadata relevant to the sub-document update operation.

Reviewer: Tighten up wording to make it more procedural and less descriptive.

```suggestion
. Call the `mutateIn()` method, which takes a document ID and an array of `MutateInSpec` objects. 

. Use a `MutateInSpec` object to specify the sub-operation to be performed within the lookup.

A `MutateInResult` object is returned, containing the result and metadata relevant to the sub-document update operation.
```

{kv-subdoc-delete-example}

[source,java]
----
include::java-sdk:hello-world:example$KvHelloWorldScoped.java[tag=kv-remove-subdoc,indent=0]
----

NOTE: If the path doesn't exist, the SDK will return a `PathNotFoundException` error.

{github}
--

Node.js::
+ 
--

> . To remove a specific field inside a document, call the `mutateIn()` method, which takes a document ID and an array of `MutateInSpec` objects. `MutateInSpec` represents the sub-operation to be performed within the lookup.
> 
> . A `MutateInResult` object is returned containing the result and metadata relevant to the sub-document update operation.

Reviewer: Tighten up wording to make it more procedural and less descriptive.

```suggestion
. Call the `mutateIn()` method, which takes a document ID and an array of `MutateInSpec` objects.

. Use a `MutateInSpec` object to specify the sub-operation to be performed within the lookup.

A `MutateInResult` object is returned, containing the result and metadata relevant to the sub-document update operation.
```

{kv-subdoc-delete-example}

[source,nodejs]
----
include::nodejs-sdk:hello-world:example$kv-hello-world-scoped.js[tag=kv-remove-subdoc,indent=0]
----

NOTE: If the path doesn't exist, the SDK will return a `PathNotFoundError` error.

{github}
--

Python::
+ 
--

> . To remove a specific field inside a document, call the `lookup_in()` function, which takes a document ID and a list of `MutateInSpec` objects. 
> `MutateInSpec` represents the sub-operation to be performed within the lookup.
> 
> . A `MutateInResult` object is returned containing the result and metadata relevant to the sub-document get operation.

Reviewer: Tighten up wording to make it more procedural and less descriptive.

```suggestion
. Call the `lookup_in()` function, which takes a document ID and a list of `MutateInSpec` objects.

. Use a `MutateInSpec` object to specify the sub-operation to be performed within the lookup.

A `MutateInResult` object is returned, containing the result and metadata relevant to the sub-document get operation.
```

{github}

For further details, refer to {java-api-url}/Collection.html[Collection^].
--

Node.js::
+ 
--
Use the `remove()` function to delete a document from the database.

'''

{kv-delete-example}

[source,nodejs]
----
include::nodejs-sdk:hello-world:example$kv-hello-world-scoped.js[tag=kv-remove,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundError` error.

{github}

For further details, refer to {nodejs-api-url}/Collection.html[Collection^].
--

Python::
+ 
--
Use the `remove()` function to delete a document from the database.

'''

{kv-delete-example}

[source,python]
----
include::python-sdk:hello-world:example$kv_hello_world_scoped.py[tag=kv-remove,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundException` error.

{github}

For further details, refer to {python-api-url}#collection-object[Collection^].
--
====

== Deleting a Sub-Document

To delete a specific field within a document you can perform a _Sub-Document_ remove operation.

[{tabs}]
====
cbc-subdoc::
+ 
--
. Connect to the `cbc-subdoc` interactive shell.

. Use the `remove` command to delete a field from a document.

> .	Pass the field to remove with the `--path` argument.

Reviewer: Not sure about the tab character!

```suggestion
. Pass the field to remove with the `--path` argument.
```
